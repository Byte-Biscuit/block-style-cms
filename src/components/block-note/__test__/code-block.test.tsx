import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import { useTheme } from "next-themes";
import { useTranslations } from "next-intl";
import CodeBlock from "../code-block";

// Mock dependencies
vi.mock("next-themes", () => ({
    useTheme: vi.fn(),
}));

vi.mock("next-intl", () => ({
    useTranslations: vi.fn(),
}));

vi.mock("@/components/code-loading", () => ({
    default: ({ variant, lines }: { variant: string; lines: number }) => (
        <div
            data-testid="code-loading"
            data-variant={variant}
            data-lines={lines}
        >
            Loading code...
        </div>
    ),
}));

// Mock react-syntax-highlighter with Prism
vi.mock("react-syntax-highlighter", () => ({
    Prism: ({
        language,
        children,
        showLineNumbers,
    }: {
        language: string;
        children: string;
        showLineNumbers?: boolean;
        customStyle?: unknown;
    }) => (
        <pre
            data-testid="syntax-highlighter"
            data-language={language}
            data-show-line-numbers={showLineNumbers}
            className="syntax-highlighter"
        >
            <code>{children}</code>
        </pre>
    ),
}));

describe("CodeBlock", () => {
    const mockUseTheme = vi.mocked(useTheme);
    const mockUseTranslations = vi.mocked(useTranslations);

    beforeEach(() => {
        mockUseTheme.mockReturnValue({
            resolvedTheme: "light",
            themes: ["light", "dark"],
            setTheme: vi.fn(),
            theme: "light",
            systemTheme: "light",
            forcedTheme: undefined,
        });

        // Mock useTranslations to return a simple translator function
        mockUseTranslations.mockImplementation(() => {
            const translations: Record<string, string> = {
                copied: "Copied!",
                copy: "Copy",
            };

            return ((key: string) => translations[key] ?? key) as ReturnType<
                typeof useTranslations
            >;
        });
    });

    it("renders code block with vitest test code (from data sample)", async () => {
        const vitestCode = `import { expect, test } from 'vitest'

test('snapshot', () => {
  // in Jest and Vitest
  expect(new Error('error')).toMatchInlineSnapshot(\`[Error: error]\`)

  // Jest snapshots \`Error.message\` for \`Error\` instance
  // Vitest prints the same value as toMatchInlineSnapshot
  expect(() => {
    throw new Error('error')
  }).toThrowErrorMatchingInlineSnapshot(\`[Error: error]\`) 
})`;

        const { asFragment } = render(
            <CodeBlock language="" code={vitestCode} />
        );

        // Wait for dynamic import to resolve
        await waitFor(() => {
            const syntaxHighlighter =
                screen.queryByTestId("syntax-highlighter");
            expect(syntaxHighlighter).toBeTruthy();
        });

        const syntaxHighlighter = screen.getByTestId("syntax-highlighter");
        expect(syntaxHighlighter).toBeTruthy();
        expect(syntaxHighlighter.getAttribute("data-language")).toBe("");

        // Verify the code content is rendered
        const codeElement = syntaxHighlighter.querySelector("code");
        expect(codeElement).toBeTruthy();
        expect(codeElement?.textContent).toBe(vitestCode);

        // Verify copy button is present
        const copyButton = screen.getByRole("button", { name: /copy/i });
        expect(copyButton).toBeTruthy();
        expect(copyButton.textContent).toContain("Copy");

        // Snapshot test
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders code block with typescript language", async () => {
        const tsCode = `interface User {
  id: number;
  name: string;
}`;

        const { asFragment } = render(
            <CodeBlock language="ts" code={tsCode} />
        );

        await waitFor(() => {
            const syntaxHighlighter =
                screen.queryByTestId("syntax-highlighter");
            expect(syntaxHighlighter).toBeTruthy();
        });

        const syntaxHighlighter = screen.getByTestId("syntax-highlighter");
        //expect(syntaxHighlighter.getAttribute("data-language")).toBe("typescript");

        // Verify language tag shows "TS"
        const languageTag = screen.getByText("TS");
        expect(languageTag).toBeTruthy();

        const codeElement = syntaxHighlighter.querySelector("code");
        expect(codeElement?.textContent).toBe(tsCode);
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders code block with custom language", async () => {
        const pythonCode = `def hello():
    print("Hello World!")`;

        const { asFragment } = render(
            <CodeBlock language="python" code={pythonCode} />
        );

        await waitFor(() => {
            const syntaxHighlighter =
                screen.queryByTestId("syntax-highlighter");
            expect(syntaxHighlighter).toBeTruthy();
        });

        const syntaxHighlighter = screen.getByTestId("syntax-highlighter");
        expect(syntaxHighlighter.getAttribute("data-language")).toBe("python");

        // Verify language tag shows "PYTHON"
        const languageTag = screen.getByText("PYTHON");
        expect(languageTag).toBeTruthy();
        expect(asFragment()).toMatchSnapshot();
    });
});
