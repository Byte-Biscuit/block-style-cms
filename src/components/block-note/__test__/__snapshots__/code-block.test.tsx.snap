// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CodeBlock > renders code block with custom language 1`] = `
<DocumentFragment>
  <figure
    aria-label="Code sample in python"
    class="relative my-4 w-full"
    data-language="python"
  >
    <div
      class="absolute top-2 right-2 z-10 flex items-center gap-2"
    >
      <span
        aria-hidden="true"
        class="inline-flex items-center rounded-md border border-gray-200 bg-gray-100 px-2 py-0.5 text-xs font-medium text-gray-700 dark:border-gray-700 dark:bg-gray-800 dark:text-gray-200"
      >
        PYTHON
      </span>
      <button
        aria-label="Copy python code"
        class="rounded-md bg-gray-700 px-3 py-1 text-xs text-white transition hover:bg-gray-600"
        type="button"
      >
        Copy
      </button>
    </div>
    <pre
      class="syntax-highlighter"
      data-language="python"
      data-show-line-numbers="true"
      data-testid="syntax-highlighter"
    >
      <code>
        def hello():
    print("Hello World!")
      </code>
    </pre>
  </figure>
</DocumentFragment>
`;

exports[`CodeBlock > renders code block with typescript language 1`] = `
<DocumentFragment>
  <figure
    aria-label="Code sample in ts"
    class="relative my-4 w-full"
    data-language="ts"
  >
    <div
      class="absolute top-2 right-2 z-10 flex items-center gap-2"
    >
      <span
        aria-hidden="true"
        class="inline-flex items-center rounded-md border border-gray-200 bg-gray-100 px-2 py-0.5 text-xs font-medium text-gray-700 dark:border-gray-700 dark:bg-gray-800 dark:text-gray-200"
      >
        TS
      </span>
      <button
        aria-label="Copy ts code"
        class="rounded-md bg-gray-700 px-3 py-1 text-xs text-white transition hover:bg-gray-600"
        type="button"
      >
        Copy
      </button>
    </div>
    <pre
      class="syntax-highlighter"
      data-language="ts"
      data-show-line-numbers="true"
      data-testid="syntax-highlighter"
    >
      <code>
        interface User {
  id: number;
  name: string;
}
      </code>
    </pre>
  </figure>
</DocumentFragment>
`;

exports[`CodeBlock > renders code block with vitest test code (from data sample) 1`] = `
<DocumentFragment>
  <figure
    aria-label="Code sample in "
    class="relative my-4 w-full"
    data-language=""
  >
    <div
      class="absolute top-2 right-2 z-10 flex items-center gap-2"
    >
      <button
        aria-label="Copy  code"
        class="rounded-md bg-gray-700 px-3 py-1 text-xs text-white transition hover:bg-gray-600"
        type="button"
      >
        Copy
      </button>
    </div>
    <pre
      class="syntax-highlighter"
      data-language=""
      data-show-line-numbers="true"
      data-testid="syntax-highlighter"
    >
      <code>
        import { expect, test } from 'vitest'

test('snapshot', () =&gt; {
  // in Jest and Vitest
  expect(new Error('error')).toMatchInlineSnapshot(\`[Error: error]\`)

  // Jest snapshots \`Error.message\` for \`Error\` instance
  // Vitest prints the same value as toMatchInlineSnapshot
  expect(() =&gt; {
    throw new Error('error')
  }).toThrowErrorMatchingInlineSnapshot(\`[Error: error]\`) 
})
      </code>
    </pre>
  </figure>
</DocumentFragment>
`;
